lib/services/local/presence.go

```diff
@@ -1033,6 +1033,79 @@ func (s *PresenceService) DeleteAllKubeServices(ctx context.Context) error {
 	))
 }
 
+// GetDatabaseServers returns all registered database proxy servers.
+func (s *PresenceService) GetDatabaseServers(ctx context.Context, namespace string, opts ...services.MarshalOption) ([]services.DatabaseServer, error) {
+	if namespace == "" {
+		return nil, trace.BadParameter("missing namespace")
+	}
+	startKey := backend.Key(dbServersPrefix, namespace)
+	result, err := s.GetRange(ctx, startKey, backend.RangeEnd(startKey), backend.NoLimit)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	servers := make([]services.DatabaseServer, len(result.Items))
+	for i, item := range result.Items {
+		server, err := services.GetDatabaseServerMarshaler().UnmarshalDatabaseServer(
+			item.Value,
+			services.KindDatabaseServer,
+			services.AddOptions(opts,
+				services.WithResourceID(item.ID),
+				services.WithExpires(item.Expires))...)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		servers[i] = server
+	}
+	return servers, nil
+}
+
+// UpsertDatabaseServer registers new database proxy server.
+func (s *PresenceService) UpsertDatabaseServer(ctx context.Context, server services.DatabaseServer) (*services.KeepAlive, error) {
+	if err := server.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	value, err := services.GetDatabaseServerMarshaler().MarshalDatabaseServer(server)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	// Because there may be multiple database servers on a single host,
+	// they are stored under the following path in the backend:
+	//   /databaseServers/<namespace>/<host-uuid>/<name>
+	lease, err := s.Put(ctx, backend.Item{
+		Key: backend.Key(dbServersPrefix,
+			server.GetNamespace(),
+			server.GetHostID(),
+			server.GetName()),
+		Value:   value,
+		Expires: server.Expiry(),
+		ID:      server.GetResourceID(),
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if server.Expiry().IsZero() {
+		return &services.KeepAlive{}, nil
+	}
+	return &services.KeepAlive{
+		Type:    services.KeepAlive_DATABASE,
+		LeaseID: lease.ID,
+		HostID:  server.GetHostID(),
+		Name:    server.GetName(),
+	}, nil
+}
+
+// DeleteDatabaseServer removes the specified database proxy server.
+func (s *PresenceService) DeleteDatabaseServer(ctx context.Context, namespace, hostID, name string) error {
+	key := backend.Key(dbServersPrefix, namespace, hostID, name)
```

Comment:
Validate name and hostID are not empty? Otherwise it behaves as `DeleteAllDatabaseServers` which might be unintended.